src/docs/torque.md

```diff
@@ -207,112 +178,236 @@ type int32 generates 'TNode<Int32T>' constexpr 'int32_t';
 
-Union types express that a value belongs to one of several possible types. We only allow union types for tagged values, because they can be distinguished at runtime using the map pointer. For example, JavaScript numbers are either Smi values or allocated `HeapNumber` objects.
+##### Class fields
+
+As well as plain values, as in the example above, class fields may contain indexed data. Here's an example:
 
 ```torque
-type Number = Smi | HeapNumber;
+extern class CoverageInfo extends HeapObject {
+  const slot_count: int32;
+  slots[slot_count]: CoverageInfoSlot;
+}
 ```
 
-Union types satisfy the following equalities:
+This means that instances of `CoverageInfo` are of varying sizes based on the data in `slot_count`.
 
-- `A | B = B | A`
-- `A | (B | C) = (A | B) | C`
-- `A | B = A` if `B` is a subtype of `A`
+Unlike C++, Torque will not implicitly add padding between fields; instead, it will fail and emit an error if fields are not properly aligned. Torque also requires that strong fields, weak fields, and scalar fields be together with other fields of the same category in the field order.
 
-It is only allowed to form union types from tagged types because untagged types cannot be distinguished at runtime.
+`const` means that a field cannot be altered at runtime (or at least not easily; Torque will fail compilation if you attempt to set it). This is a good idea for length fields, which should only be reset with great care because they would require freeing any released space and might cause data races with a marking thread.
 
-When mapping union types to CSA, the most specific common supertype of all the types of the union type is selected, with the exception of `Number` and `Numeric`, which are mapped to the corresponding CSA union types.
+`weak` at the beginning of a field declaration means that the field is visited with some kind of custom weak object semantics. If the object may be a normal weak ref (with the second bit set), then `Weak<T>` should be used in the type instead. As an example, this field from `Map` can contain some strong and some weak types, and also requires custom weak visitation:
+
+```torque
+  weak transitions_or_prototype_info: Map|Weak<Map>|TransitionArray|
+      PrototypeInfo|Smi;
```

Comment:
The truth is: "weak" doesn't have any proper semantics and just affects the generation of the kStartOf...-style constants. As far as I can tell from the actual BodyDescriptor, the transitions_or_prototype_info field is a normal field with tagged weak references, but without custom weakness. The reason why removing the weak keyword would not work at the moment is that the body descriptor actually says:

    IteratePointers(obj, Map::kStartOfStrongFieldsOffset,
                    Map::kEndOfStrongFieldsOffset, v);
    IterateMaybeWeakPointer(obj, kTransitionsOrPrototypeInfoOffset, v);

So we need the weak keyword to keep the Map::kEndOfStrongFieldsOffset correct.
This should go away and we should stop using these constants by using Torque-generated body descriptors, hopefully soon....